pinot-broker/src/main/java/org/apache/pinot/broker/routing/v2/instanceselector/BaseInstanceSelector.java

```diff
@@ -0,0 +1,177 @@
+import org.apache.pinot.common.utils.HashUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+abstract class BaseInstanceSelector implements InstanceSelector {
+  private static final Logger LOGGER = LoggerFactory.getLogger(BaseInstanceSelector.class);
+
+  // To prevent int overflow, reset the request id once it reaches this value
+  private static final int MAX_REQUEST_ID = 1_000_000_000;
+
+  private final AtomicLong _requestId = new AtomicLong();
+  private final String _tableNameWithType;
+  private final BrokerMetrics _brokerMetrics;
+
+  private volatile Set<String> _enabledInstances;
+  private volatile Map<String, List<String>> _segmentToInstancesMap;
+  private volatile Map<String, List<String>> _instanceToSegmentsMap;
+  private volatile Map<String, List<String>> _segmentToEnabledInstancesMap;
+
+  BaseInstanceSelector(String tableNameWithType, BrokerMetrics brokerMetrics) {
+    _tableNameWithType = tableNameWithType;
+    _brokerMetrics = brokerMetrics;
+  }
+
+  @Override
+  public void init(Set<String> enabledInstances, ExternalView externalView, Set<String> onlineSegments) {
+    _enabledInstances = enabledInstances;
+    onExternalViewChange(externalView, onlineSegments);
+  }
+
+  @Override
+  public void onInstancesChange(Set<String> enabledInstances, List<String> changedInstances) {
+    _enabledInstances = enabledInstances;
+
+    // Update all segments served by the changed instances
+    Set<String> segmentsToUpdate = new HashSet<>();
+    Map<String, List<String>> instanceToSegmentsMap = _instanceToSegmentsMap;
+    for (String instance : changedInstances) {
+      List<String> segments = instanceToSegmentsMap.get(instance);
+      if (segments != null) {
+        segmentsToUpdate.addAll(segments);
+      }
+    }
+
+    // Directly return if no segment needs to be updated
+    if (segmentsToUpdate.isEmpty()) {
+      return;
+    }
+
+    // Generate a new map from segment to enabled instances
+    Map<String, List<String>> segmentToInstancesMap = _segmentToInstancesMap;
+    Map<String, List<String>> segmentToEnabledInstancesMap = _segmentToEnabledInstancesMap;
+    Map<String, List<String>> newSegmentToEnabledInstancesMap =
+        new HashMap<>(HashUtil.getHashMapCapacity(segmentToEnabledInstancesMap.size()));
+    for (Map.Entry<String, List<String>> entry : segmentToEnabledInstancesMap.entrySet()) {
+      String segment = entry.getKey();
+      if (segmentsToUpdate.contains(segment)) {
+        newSegmentToEnabledInstancesMap.put(segment,
+            calculateEnabledInstancesForSegment(segment, segmentToInstancesMap.get(segment), enabledInstances));
+      } else {
+        newSegmentToEnabledInstancesMap.put(segment, entry.getValue());
+      }
+    }
+    _segmentToEnabledInstancesMap = newSegmentToEnabledInstancesMap;
+  }
+
+  @Override
+  public void onExternalViewChange(ExternalView externalView, Set<String> onlineSegments) {
```

Comment:
Can we add some documentation here explaining how the "onlineSegments" is used? It will empty or null if the table is offline?

Reply:
`onlineSegments` are segments with ONLINE/CONSUMING instances in ideal state. There are javadoc in the interface

Reply:
We don't need to skip segments not in the ideal state in the instance selector because they won't be selected in the instance selector. I excluded them to save the calculation for the unnecessary segments. While, maybe the overhead of the map lookup is higher, and because that is causing confusion, I just removed this check.